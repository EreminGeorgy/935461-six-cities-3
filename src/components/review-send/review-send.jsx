import React, {useRef, useCallback, useEffect} from 'react';
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {Operation} from "../../reducer/comments/comments.js";
import {getAuthorizationStatus} from "../../reducer/user/selectors.js";
import {AuthorizationStatus} from "../../reducer/user/user";
import {getCommentState} from "../../reducer/comments/selectors.js";
import {CommentsOperationStatus} from "../../reducer/comments/comments";

const MAX_LENGHT = `300`;
const MIN_LENGTH = `50`;

export const ReviewSend = ({sendReview, sendReviewStatus, authorizationStatus, id}) => {

  const commentRef = useRef(null);
  const submitRef = useRef(null);
  let hotelRating = null;

  const blockSubmit = () => {
    submitRef.current.disabled = true;
  };

  const unblockSubmit = () => {
    submitRef.current.disabled = false;
  };

  const onInput = () => {

    if (hotelRating) {
      unblockSubmit();
    }
  };

  const changeRatingHandler = (evt) => {
    hotelRating = parseInt(evt.target.value, 10);
    onInput();
  };

  const onSubmit = (evt) => {

    evt.preventDefault();
    blockSubmit();

    sendReview({
      comment: commentRef.current.value,
      rating: hotelRating,
      id,
    });
    evt.target.reset();
  };

  useEffect(() => {
    if (!submitRef.current) {
      return;
    }
    blockSubmit();
  }, []);


  useEffect(() => {
    if (sendReviewStatus === CommentsOperationStatus.ERROR) {
      unblockSubmit();
    }
  }, [sendReviewStatus]);

  const memoSubmit = useCallback(onSubmit, []);

  return (authorizationStatus === AuthorizationStatus.AUTH) ? (
    <form className="reviews__form form" action="#" method="post" onSubmit={memoSubmit}>
      <label className="reviews__label form__label" htmlFor="review">Your review</label>
      <div className="reviews__rating-form form__rating">
        <input className="form__rating-input visually-hidden" name="rating" value="5" id="5-stars" type="radio" onClick={changeRatingHandler} />
        <label htmlFor="5-stars" className="reviews__rating-label form__rating-label" title="perfect">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input className="form__rating-input visually-hidden" name="rating" value="4" id="4-stars" type="radio" onClick={changeRatingHandler}/>
        <label htmlFor="4-stars" className="reviews__rating-label form__rating-label" title="good">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input className="form__rating-input visually-hidden" name="rating" value="3" id="3-stars" type="radio" onClick={changeRatingHandler}/>
        <label htmlFor="3-stars" className="reviews__rating-label form__rating-label" title="not bad">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input className="form__rating-input visually-hidden" name="rating" value="2" id="2-stars" type="radio" onClick={changeRatingHandler}/>
        <label htmlFor="2-stars" className="reviews__rating-label form__rating-label" title="badly">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input className="form__rating-input visually-hidden" name="rating" value="1" id="1-star" type="radio" onClick={changeRatingHandler}/>
        <label htmlFor="1-star" className="reviews__rating-label form__rating-label" title="terribly">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>
      </div>
      <textarea className="reviews__textarea form__textarea" onChange={onInput} id="review" name="review" maxLength={MAX_LENGHT} minLength={MIN_LENGTH} placeholder="Tell how was your stay, what you like and what can be improved" ref={commentRef}></textarea>
      <div className="reviews__button-wrapper">
        <p className="reviews__help">
          To submit review please make sure to set <span className="reviews__star">rating</span> and describe your stay with at least <b className="reviews__text-amount">50 characters</b>.
        </p>
        <button className="reviews__submit form__submit button" type="submit" ref={submitRef} disabled={sendReviewStatus === CommentsOperationStatus.SUCCESS ? false : true}>Submit</button>
      </div>
      <p>{(sendReviewStatus === CommentsOperationStatus.ERROR) ? `что-то пошло не так` : ``}</p>
    </form>) : ``;
};

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  sendReviewStatus: getCommentState(state),
});

const mapDispatchToProps = (dispatch) => ({
  sendReview(reviewData) {
    dispatch(Operation.sendComment(reviewData));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ReviewSend);

ReviewSend.propTypes = {
  sendReview: PropTypes.func,
  authorizationStatus: PropTypes.string,
  sendReviewStatus: PropTypes.string,
  id: PropTypes.number,
};
