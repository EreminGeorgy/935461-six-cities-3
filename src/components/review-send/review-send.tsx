import * as React from 'react';
import {useRef, useEffect} from 'react';
import {connect} from "react-redux";
import {Operation} from "../../reducer/comments/comments";
import {getAuthorizationStatus} from "../../reducer/user/selectors";
import {AuthorizationStatus} from "../../reducer/user/user";
import {getCommentState} from "../../reducer/comments/selectors";
import {CommentsOperationStatus} from "../../reducer/comments/comments";

const MAX_LENGHT = 300;
const MIN_LENGTH = 50;

interface IProps {
  sendReview?: ({comment, rating, id}: {comment: string; rating: number; id: number}) => void;
  sendReviewStatus?: string;
  authorizationStatus?: string;
  id: number;
}

export const ReviewSend:React.FC<IProps> = (props) => {
  const {sendReview, sendReviewStatus, authorizationStatus, id} = props;
  const commentRef = useRef(null);
  const submitRef = useRef(null);
  const formRef = useRef(null);
  let hotelRating = null;

  const blockSubmit = () => {
    submitRef.current.disabled = true;
  };

  const unblockSubmit = () => {
    submitRef.current.disabled = false;
  };

  const onInput = () => {
    if (hotelRating && commentRef.current.value) {
      unblockSubmit();
    }
  };

  const changeRatingHandler = (evt) => {
    if (commentRef.current.value) {
      unblockSubmit();
    }
    hotelRating = parseInt(evt.target.value, 10);
  };

  const onSubmit = (evt) => {
    evt.preventDefault();
    blockSubmit();

    sendReview({
      comment: commentRef.current.value,
      rating: hotelRating,
      id,
    });
  };

  useEffect(() => {
    if (sendReviewStatus === CommentsOperationStatus.ERROR) {
      unblockSubmit();
    }
    if (sendReviewStatus === CommentsOperationStatus.SUCCESS) {
      if (formRef.current) {
        formRef.current.reset();
      }
    }
  }, [sendReviewStatus]);

  return (authorizationStatus === AuthorizationStatus.AUTH) ? (
    <form className="reviews__form form" action="#" method="post" onSubmit={onSubmit} ref={formRef}>
      <label className="reviews__label form__label" htmlFor="review">Your review</label>
      <div className="reviews__rating-form form__rating">
        <input className="form__rating-input visually-hidden" name="rating" value="5" id="5-stars" type="radio" onClick={changeRatingHandler} />
        <label htmlFor="5-stars" className="reviews__rating-label form__rating-label" title="perfect">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input className="form__rating-input visually-hidden" name="rating" value="4" id="4-stars" type="radio" onClick={changeRatingHandler}/>
        <label htmlFor="4-stars" className="reviews__rating-label form__rating-label" title="good">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input className="form__rating-input visually-hidden" name="rating" value="3" id="3-stars" type="radio" onClick={changeRatingHandler}/>
        <label htmlFor="3-stars" className="reviews__rating-label form__rating-label" title="not bad">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input className="form__rating-input visually-hidden" name="rating" value="2" id="2-stars" type="radio" onClick={changeRatingHandler}/>
        <label htmlFor="2-stars" className="reviews__rating-label form__rating-label" title="badly">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>

        <input className="form__rating-input visually-hidden" name="rating" value="1" id="1-star" type="radio" onClick={changeRatingHandler}/>
        <label htmlFor="1-star" className="reviews__rating-label form__rating-label" title="terribly">
          <svg className="form__star-image" width="37" height="33">
            <use xlinkHref="#icon-star"></use>
          </svg>
        </label>
      </div>
      <textarea className="reviews__textarea form__textarea" onChange={onInput} id="review" name="review" maxLength={MAX_LENGHT} minLength={MIN_LENGTH} required placeholder="Tell how was your stay, what you like and what can be improved" ref={commentRef}></textarea>
      <div className="reviews__button-wrapper">
        <p className="reviews__help">
          To submit review please make sure to set <span className="reviews__star">rating</span> and describe your stay with at least <b className="reviews__text-amount">50 characters</b>.
        </p>
        <button className="reviews__submit form__submit button" type="submit" ref={submitRef} disabled>Submit</button>
      </div>
      <p style={{color: `red`}}>{(sendReviewStatus === CommentsOperationStatus.ERROR) ? `Oooops, something went wrong!` : ``}</p>
    </form>) : <></>;
};

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  sendReviewStatus: getCommentState(state),
});

const mapDispatchToProps = (dispatch) => ({
  sendReview(reviewData) {
    dispatch(Operation.sendComment(reviewData));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ReviewSend);
