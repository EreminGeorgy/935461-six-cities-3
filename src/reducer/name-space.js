const NameSpace = {
  DATA: `DATA`,
  APPLICATION: `APPLICATION`,
  USER: `USER`,
  COMMENTS: `COMMENTS`,
  FAVORITES: `FAVORITES`,
};

export default NameSpace;
