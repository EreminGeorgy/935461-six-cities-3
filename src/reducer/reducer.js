import {offers} from '../mocks/offers';
import {extend, getCities, getOffersByCity} from '../utils/utils.js';

const citiesList = getCities(offers);
const initialOffers = getOffersByCity(offers, citiesList[0]);

const initialState = {
  activeCity: citiesList[0],
  offersInActiveCity: initialOffers,
  cities: citiesList,
};

const ActionCreator = {
  newCity: (newCity) => ({
    type: `NEW_CITY`,
    payload: newCity,
  }),
  getOffers: (newCity) => ({
    type: `GET_OFFERS`,
    payload: newCity,
  }),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case `NEW_CITY`:
      return extend(state, {
        activeCity: action.payload,
      });
    case `GET_OFFERS`:
      return extend(state, {
        offersInActiveCity: getOffersByCity(offers, action.payload),
      });
  }
  return state;
};

export {reducer, ActionCreator};
