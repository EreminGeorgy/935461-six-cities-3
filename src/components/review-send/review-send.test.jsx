import React from "react";
import renderer from "react-test-renderer";
import {ReviewSend} from "./review-send.jsx";
import reducer from '../../reducer/reducer.js';
import {createStore} from 'redux';
import {Provider} from 'react-redux';

const store = createStore(
    reducer
);

it(`<ReviewSend /> should render ReviewSend`, () => {
  const tree = renderer
    .create(
        <Provider store={store}>
          <ReviewSend/>
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
