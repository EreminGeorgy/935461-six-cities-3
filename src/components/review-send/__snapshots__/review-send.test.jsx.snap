// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewSend /> should render ReviewSend 1`] = `""`;
